pmd-core/src/main/java/net/sourceforge/pmd/RuleFactory.java

```diff
@@ -0,0 +1,405 @@
+            Node node = propertiesNode.getChildNodes().item(i);
+            if (node.getNodeType() == Node.ELEMENT_NODE && PROPERTY.equals(node.getNodeName())
+                && isPropertyDefinition((Element) node)) {
+                PropertyDescriptor<?> descriptor = parsePropertyDefinition((Element) node);
+                builder.defineProperty(descriptor);
+            }
+        }
+    }
+
+
+    /**
+     * Gets a mapping of property name to its value from the given property element.
+     *
+     * @param propertyElement Property element
+     *
+     * @return An entry of property name to its value
+     */
+    private Entry<String, String> getPropertyValue(Element propertyElement) {
+        String name = propertyElement.getAttribute(PropertyDescriptorField.NAME.attributeName());
+        return new SimpleEntry<>(name, valueFrom(propertyElement));
+    }
+
+
+    /**
+     * Overrides the rule's properties with the values defined in the element.
+     *
+     * @param rule          The rule
+     * @param propertiesElt The {@literal <properties>} element
+     */
+    private void setPropertyValues(Rule rule, Element propertiesElt) {
+        Map<String, String> overridden = getPropertyValuesFrom(propertiesElt);
+
+        for (Entry<String, String> e : overridden.entrySet()) {
+            PropertyDescriptor<?> descriptor = rule.getPropertyDescriptor(e.getKey());
+            if (descriptor == null) {
+                throw new IllegalArgumentException(
+                        "Cannot set non-existent property '" + e.getKey() + "' on Rule " + rule.getName());
+            }
+
+            setRulePropertyCapture(rule, descriptor, e.getValue());
+        }
+    }
+
+
+    private <T> void setRulePropertyCapture(Rule rule, PropertyDescriptor<T> descriptor, String value) {
+        rule.setProperty(descriptor, descriptor.valueFrom(value));
+    }
+
+
+    /**
+     * Finds out if the property element defines a property.
+     *
+     * @param node Property element
+     *
+     * @return True if this element defines a new property, false if this is just stating a value
+     */
+    private static boolean isPropertyDefinition(Element node) {
+        return StringUtils.isNotBlank(node.getAttribute(PropertyDescriptorField.TYPE.attributeName()));
+    }
+
+
+    /**
+     * Parses a property definition node and returns the defined property descriptor.
+     *
+     * @param propertyElement Property node to parse
+     *
+     * @return The property descriptor
+     */
+    private static PropertyDescriptor<?> parsePropertyDefinition(Element propertyElement) {
+
+        String typeId = propertyElement.getAttribute(PropertyDescriptorField.TYPE.attributeName());
+
+        PropertyDescriptorExternalBuilder<?> pdFactory = PropertyDescriptorUtil.factoryFor(typeId);
+        if (pdFactory == null) {
+            throw new RuntimeException("No property descriptor factory for type: " + typeId);
```

Comment:
why are other configuration errors `IllegarlArgumentExceptions` but this one is a generic `RuntimeException`?